cobbler/actions/buildiso/netboot.py

```diff
@@ -246,7 +248,7 @@ def _generate_append_suse(self):
             del self.data["kernel_options"]["install"]
         else:
             self.append_line += (
-                f" install=http://{self.data['server']}:{self.data['http_port']}/cblr/"
+                f" install={protocol}://{self.data['server']}:{self.data['http_port']}/cblr/"
                 f"links/{self.dist.name}"
```

Comment:
The problem is, on first look, that `self.dist` can be `None`.

Reply:
First I thought "Well that can't be" but on a more close look it seems you are right. I think since the distro name and the distro object should never differentiate from each another, we should be able to safely merge them because the distro object will of course have the name attached.

Reply:
Okay well I was wrong in terms of the distro name. We can't replace that access (`make_sorter`), however we still will be able to ensure that the distro object is properly set with a small refactoring.

Reply:
Okay programming when one can't sleep does not yield in the best results... Sorry for the back and forth but now a definitive answer:

- `self.dist` in the `AppendLineBuilder` is only used in the `generate_system` case, `generate_profile` doesn't make use of this property or in fact any of the other methods in the class.
- `self.dist` is being set at the beginning of `generate_system`.
- In theory `dist` can be supplied with `None` and we indeed don't do a check for `None`. However I don't see a case where we can pass `None` as in all cases we retrieve the dist via the system object and thus by definition of a valid system it's parent distro can never be `None`. Cobbler should refuse to start if the parent distro of a system is `None`.

Please correct me if I am wrong here but I don't see how this can be `None`.

Reply:
It's possible that it can't end up being `None` at runtime, I didn't check how to get here. I mainly commented because I don't think `mypy` is necessarily wrong. `pyright` also shows me that error. In fact, `pyright` shows me more errors that just this one for `netboot.py`:
<details><summary>Error list</summary>

```
 netboot.py    16   1 warning  too-many-instance-attributes Too many instance attributes (9/7) (python-pylint)
 netboot.py    38  26 error           Pyright [reportOptionalMemberAccess]: "breed" is not a known member of "None" (eglot)
 netboot.py    43  28 error           Pyright [reportOptionalMemberAccess]: "breed" is not a known member of "None" (eglot)
 netboot.py    50  28 error           Pyright [reportOptionalMemberAccess]: "breed" is not a known member of "None" (eglot)
 netboot.py    58  26 error           Pyright [reportOptionalMemberAccess]: "breed" is not a known member of "None" (eglot)
 netboot.py    60  28 error           Pyright [reportOptionalMemberAccess]: "breed" is not a known member of "None" (eglot)
 netboot.py    62  28 error           Pyright [reportOptionalMemberAccess]: "breed" is not a known member of "None" (eglot)
 netboot.py    70  26 error           Pyright [reportOptionalMemberAccess]: "breed" is not a known member of "None" (eglot)
 netboot.py    72  28 error           Pyright [reportOptionalMemberAccess]: "breed" is not a known member of "None" (eglot)
 netboot.py    80  26 error           Pyright [reportOptionalMemberAccess]: "breed" is not a known member of "None" (eglot)
 netboot.py    82  28 error           Pyright [reportOptionalMemberAccess]: "breed" is not a known member of "None" (eglot)
 netboot.py    91  26 error           Pyright [reportOptionalMemberAccess]: "breed" is not a known member of "None" (eglot)
 netboot.py    93  28 error           Pyright [reportOptionalMemberAccess]: "breed" is not a known member of "None" (eglot)
 netboot.py    95  28 error           Pyright [reportOptionalMemberAccess]: "breed" is not a known member of "None" (eglot)
 netboot.py   111  22 error           Pyright [reportOptionalMemberAccess]: "breed" is not a known member of "None" (eglot)
 netboot.py   117  24 error           Pyright [reportOptionalMemberAccess]: "breed" is not a known member of "None" (eglot)
 netboot.py   121  24 error           Pyright [reportOptionalMemberAccess]: "breed" is not a known member of "None" (eglot)
 netboot.py   132  22 error           Pyright [reportOptionalMemberAccess]: "breed" is not a known member of "None" (eglot)
 netboot.py   136  24 error           Pyright [reportOptionalMemberAccess]: "breed" is not a known member of "None" (eglot)
 netboot.py   140  24 error           Pyright [reportOptionalMemberAccess]: "breed" is not a known member of "None" (eglot)
 netboot.py   149  22 error           Pyright [reportOptionalMemberAccess]: "breed" is not a known member of "None" (eglot)
 netboot.py   153  24 error           Pyright [reportOptionalMemberAccess]: "breed" is not a known member of "None" (eglot)
 netboot.py   162  22 error           Pyright [reportOptionalMemberAccess]: "breed" is not a known member of "None" (eglot)
 netboot.py   166  24 error           Pyright [reportOptionalMemberAccess]: "breed" is not a known member of "None" (eglot)
 netboot.py   175  22 error           Pyright [reportOptionalMemberAccess]: "breed" is not a known member of "None" (eglot)
 netboot.py   179  24 error           Pyright [reportOptionalMemberAccess]: "breed" is not a known member of "None" (eglot)
 netboot.py   183  24 error           Pyright [reportOptionalMemberAccess]: "breed" is not a known member of "None" (eglot)
 netboot.py   235  49 error           Pyright [reportOptionalMemberAccess]: "os_version" is not a known member of "None" (eglot)
 netboot.py   252  36 error           Pyright [reportOptionalMemberAccess]: "name" is not a known member of "None" (eglot)
 netboot.py   472  41 error           Pyright [reportOptionalSubscript]: Object of type "None" is not subscriptable (eglot)
 netboot.py   473  13 error           Pyright [reportOptionalSubscript]: Object of type "None" is not subscriptable (eglot)
 netboot.py   474  43 error           Pyright [reportOptionalSubscript]: Object of type "None" is not subscriptable (eglot)
 netboot.py   474  60 error           Pyright [reportOptionalSubscript]: Object of type "None" is not subscriptable (eglot)
 netboot.py   478  71 error           Pyright [reportGeneralTypeIssues]: Argument of type "dict[Unknown, Unknown] | None" cannot be assigned to parameter "data" of type "dict[Unknown, Unknown]" in function "__init__"
   Type "dict[Unknown, Unknown] | None" cannot be assigned to type "dict[Unknown, Unknown]"
     Type "None" cannot be assigned to type "dict[Unknown, Unknown]" (eglot)
 netboot.py   505  17 error           Pyright [reportOptionalSubscript]: Object of type "None" is not subscriptable (eglot)
 netboot.py   508  41 error           Pyright [reportOptionalSubscript]: Object of type "None" is not subscriptable (eglot)
 netboot.py   510  13 error           Pyright [reportOptionalSubscript]: Object of type "None" is not subscriptable (eglot)
 netboot.py   511  43 error           Pyright [reportOptionalSubscript]: Object of type "None" is not subscriptable (eglot)
 netboot.py   511  60 error           Pyright [reportOptionalSubscript]: Object of type "None" is not subscriptable (eglot)
 netboot.py   515  71 error           Pyright [reportGeneralTypeIssues]: Argument of type "dict[Unknown, Unknown] | None" cannot be assigned to parameter "data" of type "dict[Unknown, Unknown]" in function "__init__"
   Type "dict[Unknown, Unknown] | None" cannot be assigned to type "dict[Unknown, Unknown]"
     Type "None" cannot be assigned to type "dict[Unknown, Unknown]" (eglot)
 netboot.py   545   5 warning  too-many-arguments Too many arguments (8/5) (python-pylint)
```
</details>

This particular member can be type annotated as a class variable without a default. Functionally it shouldn't change anything as far as I can tell:

```diff
modified   cobbler/actions/buildiso/netboot.py
@@ -8,6 +8,7 @@ import os
 import re
 from typing import Optional, List
 
 from cobbler import utils
+from cobbler.items import distro
 from cobbler.utils import input_converters
 from cobbler.actions import buildiso
@@ -17,12 +18,12 @@ class AppendLineBuilder:
     """
     This class is meant to be initiated for a single append line. Afterwards the object should be disposed.
     """
+    dist: distro.Distro
 
     def __init__(self, distro_name: str, data: dict):
         self.append_line = ""
         self.data = data
         self.distro_name = distro_name
-        self.dist = None
         self.system_interface = None
         self.system_ip = None
         self.system_netmask = None
```

Looks like a limitation of static analysis tools to me.

Reply:
I am aware that it can be annotated. However due to import cycles I try to prevent adding an import to a module that does not already use it. It is a bit sad indeed that the tools are not able to resolve this but it is what it is.